lib/realtime/tenants/authorization.ex

```diff
@@ -0,0 +1,82 @@
+defmodule Realtime.Tenants.Authorization do
+  @moduledoc """
+  Runs validations based on RLS policies to set permissions for a given connection.
+
+  It will assign the a new key to a socket or a conn with the following:
+  * read - a boolean indicating whether the connection has read permissions
+  """
+  require Logger
+
+  @type params :: %{
+          :channel_name => binary() | nil,
+          :claims => map(),
+          :headers => keyword({binary(), binary()}),
+          :jwt => map(),
+          :role => binary()
+        }
+
+  @spec get_authorizations(Phoenix.Socket.t() | Plug.Conn.t(), DBConnection.t(), params()) ::
+          {:ok, Phoenix.Socket.t() | Plug.Conn.t()} | {:error, :unauthorized}
+  @doc """
+  Runs validations based on RLS policies to set permissions for a given connection (either Phoenix.Socket or Plug.Conn).
+  """
+  def get_authorizations(%Phoenix.Socket{} = socket, db_conn, params) do
+    case get_permissions_for_connection(db_conn, params) do
+      {:ok, permissions} -> {:ok, Phoenix.Socket.assign(socket, :permissions, permissions)}
+      _ -> {:error, :unauthorized}
+    end
+  end
+
+  def get_authorizations(%Plug.Conn{} = conn, db_conn, params) do
+    case get_permissions_for_connection(db_conn, params) do
+      {:ok, permissions} -> {:ok, Plug.Conn.assign(conn, :permissions, permissions)}
+      _ -> {:error, :unauthorized}
+    end
+  end
+
+  defp get_permissions_for_connection(conn, params) do
+    %{
+      channel_name: channel_name,
+      headers: headers,
+      jwt: jwt,
+      claims: claims,
+      role: role
+    } = params
+
+    sub = Map.get(claims, :sub)
+    claims = Jason.encode!(claims)
+    headers = headers |> Map.new() |> Jason.encode!()
+
+    Postgrex.transaction(conn, fn conn ->
+      Postgrex.query(
+        conn,
+        """
+        SELECT
+         set_config('role', $1, true),
+         set_config('realtime.channel_name', $2, true),
+         set_config('request.jwt.claim.role', $3, true),
+         set_config('request.jwt', $4, true),
+         set_config('request.jwt.claim.sub', $5, true),
+         set_config('request.jwt.claims', $6, true),
+         set_config('request.headers', $7, true)
+        """,
+        [role, channel_name, role, jwt, sub, claims, headers]
+      )
+
+      case Postgrex.query(conn, "SELECT name from realtime.channels", [], mode: :savepoint) do
```

Comment:
I think we need to start handling exceptions from Postgrex.query and turn them into an error tuple.

Reply:
And what is `mode: :savepoint` doing for us here?

Reply:
`mode: :savepoint` avoids invalidating the transaction fully and reverts to a previous stage of the transaction which in this case allows us to have a `false` and keeping the transaction ok

```
 In case of an exception, the transaction will rollback to the savepoint and prevent the transaction from failing.
```

from https://hexdocs.pm/ecto/Ecto.Repo.html 

so this should handle part of the exceptions but we should still be careful as we are not doing that when settings configs so some exception handling needs to happen